core/src/main/java/org/apache/accumulo/core/file/blockfile/cache/lru/CachedBlockQueue.java

```diff
@@ -91,24 +93,19 @@ public void add(CachedBlock cb) {
    * @return list of cached elements in descending order
    */
   public CachedBlock[] get() {
-    LinkedList<CachedBlock> blocks = new LinkedList<>();
-    while (!queue.isEmpty()) {
-      blocks.addFirst(queue.poll());
-    }
-    return blocks.toArray(new CachedBlock[blocks.size()]);
+     CachedBlock[] blocks = queue.toArray(new CachedBlock[0]);
+     Arrays.sort(blocks, Collections.reverseOrder(queue.comparator()));
```

Comment:
This `O(n log(n))` sort function is probably more expensive than the `O(log(n))` `poll()` method `n` times. First, as `n` decreases, `log(n)` is faster and faster for `poll()`, and we can also avoid all calls to the comparator by preserving the order directly from the PriorityQueue.

So, I'm not sure that this is better than the existing code on performance measures. Have you done any tests to see whether your proposed change is faster?

It's also not thread-safe (though I haven't checked thoroughly to see if it needs to be). If data is inserted into the queue in the old code, we'd get it by calling `poll()`, but if data is inserted in the new code, we could end up missing it and clearing it. This may not matter if it doesn't need to be thread-safe... as I said, I haven't checked.

Reply:
I imagine having to call into the `poll()` method repeatedly is going to be slower.  Faster to just call a single method then sort the array instead of having to call down into the `poll()` method for each item.  I can test this out though and I look forward to it just from a personal curiosity.


I looked at the thread-safeness.  `PriorityQueue` itself is not thread safe, so I assumed no harm here.

> Note that this implementation is not synchronized. Multiple threads should not access a PriorityQueue instance concurrently if any of the threads modifies the queue. Instead, use the thread-safe PriorityBlockingQueue class.

https://docs.oracle.com/javase/7/docs/api/java/util/PriorityQueue.html

Reply:
Thanks for checking the thread-safety of PriorityQueue. I checked how we use it in LruBlockCache.java, and it looks like thread-safety is not a concern there either. So, this is just a question of efficiency. I'm inclined to prefer your code, so long as the performance isn't worse.

Reply:
```
# Run complete. Total time: 00:01:07

Benchmark                        Mode  Cnt      Score     Error  Units
BenchmarkQueueDrain.clearDrain  thrpt   30  11440.641 ± 211.193  ops/s
BenchmarkQueueDrain.whileDrain  thrpt   30  11653.781 ± 145.950  ops/s
```

So the two implementations are within the margin of error.  What I cannot measure is the cost of creating all of the nodes in the `LinkedList`.  There was no JVM pressure in my perf test so that cost was not included,... v.s. just a single array created and a sort applied.

Reply:
Hi @belugabehr . Thanks for the response. I'm going to assume that one of the above is testing the old code, and one is testing the change.

I'm looking into this a bit more, and it looks like `get()` is only being called in one place that matters (`LruBlockCache.BlockBucket.free()`), and only ever in a single-threaded scenario. This change seem to be trying to optimize constructing an ordered array so that `BlockBucket.free()` can iterate through that array. However, looking at the code, it's not clear to me why we even need this intermediate array at all, because we already have an ordered view in the priority queue itself! Couldn't we just add a `poll()` method onto `CachedBlockQueue` with a trivial implementation to call `poll()` on its internal queue, and have `free()` iterate calling that directly?

If the intermediate array can just be eliminated entirely, that might be a better change than merely preferring ArrayList over LinkedList here. Does that sound like something you'd be interested in looking into?

Reply:
@ctubbsii Nice idea.  Done.